File: Submissions/480_Sliding_Window_Median.cpp
class Solution {
public:
    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        vector<double> result;
        multiset<int> sorted;
        
        for(int i = 0; i<k; ++i){
            sorted.insert(nums[i]);
        }
        
        auto median = sorted.begin();
        std::advance(median, k/2);
        
        bool isEven = k%2 == 0;
        int size = nums.size();
        for(int i=k; ; ++i){
            if(isEven){
                auto median2 = std::prev(median, 1);
                result.push_back(((double)(*median) + *median2)/2);
            } else {
                result.push_back(*median);
            }
            
            if(i == size) break;
            
            sorted.insert(nums[i]);
            
            if(nums[i] < *median) median--;

            if(nums[i-k] <= *median) median++;
            
            auto it = sorted.find(nums[i-k]);
            sorted.erase(it);
        }
        
        return result;
    }
};

File: Kotlin/480_Sliding_Window_Median.kt
class Solution {
    fun medianSlidingWindow(nums: IntArray, k: Int): DoubleArray {
        val result = DoubleArray(nums.size-k+1)
        val sorted = mutableListOf<Double>()
        
        for(i in 0..k-1) {
            sorted.add(nums[i].toDouble())
        }
        sorted.sort()
        
        val is_even = k%2 == 0
        result[0] = if(is_even) (sorted.get(k/2-1) + sorted.get(k/2))/2.0 else sorted.get(k/2).toDouble()
        
        for(i in 0..nums.size-k-1){
            var index = sorted.binarySearch(nums[i+k].toDouble())
            if(index>=0){
                sorted.add(index, nums[i+k].toDouble());
            }else{
				sorted.add(-index-1, nums[i+k].toDouble());
            }
            index=sorted.binarySearch(nums[i].toDouble());
            sorted.removeAt(index);
            result[i+1] = if(is_even) (sorted.get(k/2-1) + sorted.get(k/2))/2.0 else sorted.get(k/2).toDouble()
            
        }
        
        return result
    }
}